core-primitives/src/lib.rs

```diff
@@ -79,20 +79,19 @@ pub use sp_runtime::OpaqueExtrinsic as UncheckedExtrinsic;
 /// the parachain.
 pub type Remark = [u8; 32];
 
-/// These are special "control" messages that can be passed from the Relaychain to a parachain.
-/// They should be handled by all parachains.
+/// A message sent from the relay-chain down to a parachain.
+///
+/// The size of the message is limited by the `config.critical_downward_message_size` parameter.
+pub type DownwardMessage = sp_std::vec::Vec<u8>;
+
+/// A wrapped version of `DownwardMessage`. The difference is that it has attached the block number when
+/// the message was sent.
 #[derive(codec::Encode, codec::Decode, Clone, sp_runtime::RuntimeDebug, PartialEq)]
-pub enum DownwardMessage<AccountId = crate::AccountId> {
-	/// Some funds were transferred into the parachain's account. The hash is the identifier that
-	/// was given with the transfer.
-	TransferInto(AccountId, Balance, Remark),
-	/// An opaque blob of data. The relay chain must somehow know how to form this so that the
-	/// destination parachain does something sensible.
-	///
-	/// NOTE: Be very careful not to allow users to place arbitrary size information in here.
-	Opaque(sp_std::vec::Vec<u8>),
-	/// XCMP message for the Parachain.
-	XCMPMessage(sp_std::vec::Vec<u8>),
+pub struct InboundDownwardMessage<BlockNumber = crate::BlockNumber> {
```

Comment:
Is `InboundDownwardMessage` named in contrast to `DownwardMessage`, or to an `OutboundDownwardMessage` (presumably defined elsewhere and used on the relay chain)?

If possible, it would be good to change these names; I would have trouble remembering which of `InboundDownwardMessage` and `DownwardMessage` I wanted. Possibly rename `DownwardMessage` => `DownwardMessageContent` and `InboundDownwardMessage` => `DownwardMessage`.

Reply:
Ok, so first HRMP: I defined two structs for HRMP messages, for inbound and outbound. The inbound is from the pov of  the receiver and the outbound is from the pov of the sender.

At that time there were only `DownwardMessage`, which is an alias to `Vec<u8>`.

But then it turned out that I needed a specific view of downward messages from pov of the receiver, specifically the one that includes the block number. So because of that, I went with the `InboundDownwardMessage` because that's a very similar case. 

Now, I left `DownwardMessage` mostly because I just didn't think about this situation. I also don't think that this naming scheme is ideal. But I don't think that the suggested naming is better for the same reason. Specifically, if I am coding some stuff that works with the downward messages, what should I take: downward message or a downward message contents? Not clear without diving into docs reading the difference between those. While `InboundDownwardMessage` suggests that it's a downward message that is used in the context of receiving.

Reply:
Moving this discussion to https://github.com/paritytech/polkadot/pull/1859